### ISSUE TITLE
Policy for uncommon formats implemented via third-party crates

### ISSUE THREAD

**Issue by Shnatsel** [image-rs/image#2363]
### The problem

There is a number of formats that are implemented as third-party crates, but are not supported by `image`:

JPEG XL: https://crates.io/crates/jxl-oxide
RAW: https://crates.io/crates/rawloader or https://crates.io/crates/quickraw
PSD subset: https://crates.io/crates/zune-psd or https://crates.io/crates/psd
ICNS: https://crates.io/crates/icns
JPEG 2000: https://crates.io/crates/openjp2 (not sure if the port is complete enough just yet)

None of these formats have specifications available publicly and for free. As per #1979, right now they are not supported by `image` at all.

This leads to each API users that needs them reimplementing the integrations into `image`.  [oculante](https://github.com/woelper/oculante) has already done this, and I'm going to have to do this in `wondermagick` too. 

### Possible solutions

Adding `image` integrations to those crates (like it's done for libwebp) is a step in the right direction, but doesn't really solve this because the API is still quite awkward: you cannot just use `reader.with_guessed_format().decode()?`, the API user has to explicitly list every format and also deal with magic numbers for them. 
Also, when `image` ships a new semver-incompatible version, every integration would have to bump their semver as well, and it could take a long while until all the format integrations are upgraded.

A solution that seems promising to me is a tiered format support, where we specifically advertise third-party crates listed above as Tier 2. We clearly state that we do not accept bug reports about them and that any issues should go to the third-party crate's issue tracker. These formats would also be disabled by default, with explicit opt-in required to enable them.
This avoids the issues with the other option, but still adds some (hopefully small) additional maintenance burden for the in-tree integration of the third-party crate.

Thoughts? Any other good options I've missed?

**Comment by fintelia** [comment]
Agreed that having tiers is a helpful frame for different levels of format support. I'd probably break up formats into more categories than just real vs. hobby though. 

* **Major formats** -- Formats like PNG, JPEG, GIF, TIFF, etc.
* **Specialized and obscure formats** -- Less common than the major formats but used because they outperform the major formats on some dimension or because they're the chosen standard in some field.
* **Emerging formats** -- Recently created formats likely to land in one of the prior categories. Frequently accompanied by a large marketing push to pressure software makers to include the format so they'll gain traction.
* **Obsolete formats** -- Before PNG was widely used, dozens of different software packages created their own lossless image formats. These are "real" formats in that they were used by real shipping software, but these days are broadly considered obsolete. They tended to either be uncompressed or only use some bespoke RLE compression scheme. IMO, in almost all cases, anyone who still has images in one of these formats should probably just convert them to PNG and move on. The description of PFM includes the line that the format came from "[a program he found somewhere](https://netpbm.sourceforge.net/doc/pfm.html)" (without any indication of what program or where it was found!), which to me really sums up how the bar was for new formats.
* **Hobby formats** -- Something Farbfeld or QOI that is relatively new but not backed by a major company or coalition. 
*  **Experimental formats** -- Stuff like FLIF or WebP 2 that aren't intended for broader use. Probably the clearest set of formats not worth supporting.

There's also enormous differences in implementation complexity between formats. Most of the obsolete and hobby formats are extremely simple.

**Comment by fintelia** [comment]
Another question is in-tree vs. out-of-tree.

I'm starting to think that we might want to make an `image-extras` crate to hold Tier 3 formats rather than accepting a large number of additional bindings and optional dependencies into `image`. Disabled-by-default formats theoretically shouldn't cause issues, but we've already had one [kind of serious bug in our Cargo.toml](https://github.com/image-rs/image/issues/2267) from a seemingly unrelated edit. A separate crate would sharply reduce the risk and if we get external hooks support right, the location of format bindings shouldn't make a big difference.

I should also add that the exact definitions of different tiers would have to be tuned. Our jpeg decoder is currently backed by `zune-jpeg`, but I don't think that should degrade it to tier 3 or even tier 2. And even `image-png` uses `simd-adler32` which is an external crate. Similarly hoping for all our enabled-by-default codecs to be fuzzed is good in principle, but at the moment there's no one interested in doing the bug triage and fixing work associated with that.

**Comment by HeroicKatora** [comment]
Does anyone feel knowledgeable enough about ffmpeg, to describe how their feature discovery mechanism works? If there is a form of dynamic loading then this is out-of-scope for now (we need _much_ better language agnostic interfaces in practice, that do not require code-loading and/or pipes to a forked process imo); but the remainder of this mechanism and their interfaces for policy would be worthwhile looking at.

**Comment by fintelia** [comment]
I'm not sure about those two libraries, but I looked into the `file` tool a while back. IIRC it has the different formats controlled via a config file that has essentially bytecode snippits for each format. Many just amount to "match the first N bytes against this constant", but it also can do fancier things like "read the next two bytes as a little endian integer and jump forward that many bytes"

**Comment by awxkee** [comment]
> Does anyone feel knowledgeable enough about ffmpeg and/or imagemagick, to describe how their format support discovery mechanism works? If there is a form of dynamic loading then this is out-of-scope for now (we need _much_ better language agnostic interfaces in practice, that do not require code-loading and/or pipes to a forked process imo); but the remainder of this mechanism and their interfaces for policy would be worthwhile looking at.

Magick have a fixed [list of methods](https://github.com/ImageMagick/ImageMagick/blob/6818700e62c36f3bb2638f04bb39235313d488f2/MagickCore/static.h#L89) that invoked at start.
If core compiled against module plugin [adds info about itself](https://github.com/ImageMagick/ImageMagick/blob/6818700e62c36f3bb2638f04bb39235313d488f2/coders/jxl.c#L816), otherwise skipping it.
Then it enumerates `Modules` that registered info about themselves, and [it is trying to load shared libs](https://github.com/ImageMagick/ImageMagick/blob/6818700e62c36f3bb2638f04bb39235313d488f2/MagickCore/module.c#L1365).
Not sure if it is helpful in Rust.

**Comment by kornelski** [comment]
Each format registers a `magic` callback function, and each format checks for itself. I've checked a few, and they don't look sophisticated, e.g. https://github.com/ImageMagick/ImageMagick/blob/main/coders/svg.c#L232

**Comment by sorairolake** [comment]
Is there a possibility to split out the low-level handling of image file formats (e.g., Farbfeld) which are implemented directly in this crate into separate crates (like `image-webp`)?

**Comment by Shnatsel** [comment]
Yes. Although most formats that are currently bundled are so simple that splitting them out didn't really make sense.

**Comment by fintelia** [comment]
I wrote up a draft policy for `image-extras` [here](https://github.com/image-rs/image-extras/blob/main/README.md). Seeing that we've recently seen interest for having PCX, XBM, XPM, WBMP, and OTB formats, I am increasingly thinking it is a good idea to have a dedicated crate for less common formats.

**Comment by fintelia** [comment]
Closing now that #2372 is merged

**Comment by 197g** [comment]
@fintelia I was also thinking we should make an issue template for format requests. Specifically, a template can be constructed as purely informational such that no issue can actually be created through it. We could then link to both `image-extra` for uncommon formats and note our policy of highly preferring rust-native crates.

Also, what do you think about changing the general strategy here, disallowing issues without links to libraries and instead pivoting to a discussion board for format requests. I think that would imply less responsibility ('please implement') than an issue, which by nature is for things that should be changed in the repository. We can still have issues for concrete crate bindings for instance, or for alternative crates that implement a format.

**Comment by fintelia** [comment]
An issue template and/or a note in the `image` crate readme seem good to me. The existing open format request issues can probably all be closed pointing at `image-extras`.

The `image-extras` repository already has a [readme section on new formats](https://github.com/image-rs/image-extras/blob/main/README.md#new-formats). I agree that we probably don't want to accept a bunch of "please implement" issues without someone interested to do the work. Though when there is implementation interest, I think either issues or discussions are fine. Overall we could probably do a better job encouraging folks to talk with us before opening PRs to check that the feature is something we want to implement and give design suggestions. Like, if a format has patent issues or whatever, I'd rather bring that up before someone has spent a bunch of time writing a PR.

**Comment by 197g** [comment]
> Overall we could probably do a better job encouraging folks to talk with us before opening PRs to check that the feature is something we want to implement and give design suggestions. Like, if a format has patent issues or whatever, I'd rather bring that up before someone has spent a bunch of time writing a PR.

Sounds good, I think we might be able to explain that better as a discussion page as well. (And maybe also make sure boxes are ticked before they get converted to issues & PRs). Overall suggesting a format is more like an RFC process but a separate repository and structure seems just a little over-the-top for now. If we notice any upside to editing and reviewing a shared _document_ we can move those discussions eventually.
